libkpod/attach.go

```diff
@@ -0,0 +1,182 @@
+	"github.com/kubernetes-incubator/cri-o/oci"
+	"github.com/kubernetes-incubator/cri-o/utils"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sys/unix"
+	"io"
+	"k8s.io/client-go/tools/remotecommand"
+	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
+	"net"
+	"os"
+	"path/filepath"
+	"strconv"
+)
+
+/* Sync with stdpipe_t in conmon.c */
+const (
+	AttachPipeStdin  = 1
+	AttachPipeStdout = 2
+	AttachPipeStderr = 3
+)
+
+// ContainerAttach attaches to a running container
+// keys are string representation of what to use to detach
+func (c *ContainerServer) ContainerAttach(container string, noStdIn bool, keys string) error {
+	// Check the validity of the provided keys first
+	// Format is a single character [a-Z] or ctrl-<value> where <value> is one of: a-z, @, ^, [, , or _.
+	var err error
+	detachKeys := []byte{}
+	if len(keys) > 0 {
+		detachKeys, err = term.ToBytes(keys)
+		if err != nil {
+			return errors.Wrapf(err, "invalid detach keys")
+		}
+	}
+
+	ctr, err := c.LookupContainer(container)
+	if err != nil {
+		return errors.Wrapf(err, "failed to find container %s", container)
+	}
+
+	cStatus := c.runtime.ContainerStatus(ctr)
+	if !(cStatus.Status == oci.ContainerStateRunning || cStatus.Status == oci.ContainerStateCreated) {
+		return errors.Errorf("%s is not created or running", container)
+	}
+
+	resize := make(chan remotecommand.TerminalSize)
+
+	AttachContainerSocket(ctr, resize, noStdIn, detachKeys)
+	close(resize)
+	c.ContainerStateToDisk(ctr)
+
+	return nil
+}
+
+// getStandardStreams returns the standard IO readers/writers
+func getStandardStreams() (stdIn, stdOut, stdErr *os.File) {
+	return os.Stdin, os.Stdout, os.Stderr
+}
+
+func closeInputStream(stdIn io.ReadCloser) {
+	stdIn.Close()
+}
+
+// AttachContainerSocket connects to the crio socket and deals with the IO
+func AttachContainerSocket(ctr *oci.Container, resize <-chan remotecommand.TerminalSize, noStdIn bool, detachKeys []byte) error {
+	inputStream, outputStream, errorStream := getStandardStreams()
+	defer closeInputStream(inputStream)
+
+	tty, err := strconv.ParseBool(ctr.State().Annotations["io.kubernetes.cri-o.TTY"])
+	if err != nil {
+		return errors.Errorf("unable to parse annotations in %", ctr.ID)
+	}
+	if !tty {
+		return errors.Errorf("no tty available", ctr.ID())
+	}
+
+	oldTermState, err := term.SaveState(inputStream.Fd())
+	defer term.RestoreTerminal(inputStream.Fd(), oldTermState)
+
+	if err != nil {
+		return errors.Wrapf(err, "unable to save terminal state")
+	}
+
+	// Put both input and output into raw
+	if !noStdIn {
+		term.SetRawTerminal(inputStream.Fd())
+	}
+
+	controlPath := filepath.Join(ctr.BundlePath(), "ctl")
+	controlFile, err := os.OpenFile(controlPath, unix.O_WRONLY, 0)
+	if err != nil {
+		return fmt.Errorf("failed to open container ctl file: %v", err)
+	}
+
+	kubecontainer.HandleResizing(resize, func(size remotecommand.TerminalSize) {
+		logrus.Infof("Got a resize event: %+v", size)
```

Comment:
Does this need to be info-level? It feels like it should be logrus.Debugf

Reply:
sure